generators/app/index.js

```diff
@@ -151,37 +155,49 @@ module.exports = Generator.extend({
           }
         );
       }
+
+      this.composer = options.drupalDistro.loadComposer(this, options);
     },
 
     // While there are write operations in this, other write operations may need
     // to examine the composer.json to determine their own configuration.
     composerJson: function () {
-      // Import any existing composer.json that was saved above.
-      var composer = {};
       var isNewProject = (this.composerOrig == undefined);
-      if (isNewProject) {
-        // Project hasn't been generated yet, so start with composer template.
-        composer = this.fs.readJSON('composer.json');
-      }
-      else {
+      if (!isNewProject) {
         // Use original composer file if project already generated.
-        composer = this.composerOrig;
+        this.composer = this.composerOrig;
       }
 
-      composer.name = options.projectName;
-      composer.description = options.projectDescription;
+      this.composer.name = options.projectName;
+      this.composer.description = options.projectDescription;
       // Allow distros to modify the composer.json.
       if (typeof options.drupalDistro.modifyComposer == 'function') {
         var done = this.async();
-        composer = options.drupalDistro.modifyComposer(this, options, composer, isNewProject, done);
+        var self = this;
+        options.drupalDistro.modifyComposer(self, options, this.composer, isNewProject, done, function(err, result, done) {
+          if (!err && result) {
+            this.composer = result;
+          }
+          done();
+        }.bind(self));
+      }
+
+      // Overwrite new project composer.json with a new core version.
+      // This needs to go down here to give distros a chance to preemptively
+      // modify the version of drupal/core.
+      if (isNewProject) {
+        // drupal/core is set, it is a real version or range and we have a
+        // release version to derive something better.
+        if (this.composer.require['drupal/core'] && !_.isString(this.composer.require['drupal/core']) && options.drupalDistroRelease) {
+          this.composer.require['drupal/core'] = require('../lib/drupalProjectVersion').toMinorRange(options.drupalDistroRelease);
```

Comment:
- [ ] Remove deferred require of drupalProjectVersion.js for a more standard global require for readability.